Add CameraFeed support for Web

shiena opened the pull request:
<!--
Please target the `master` branch in priority.

Relevant fixes are cherry-picked for stable branches as needed by maintainers.

To speed up the contribution process and avoid CI errors, please set up pre-commit hooks locally:
https://docs.godotengine.org/en/latest/contributing/development/code_style_guidelines.html
-->
fixed: https://github.com/godotengine/godot-proposals/issues/12493

Current Limitation:
~~The `platform/web/js/libs/library_godot_camera.js` library includes certain functionalities that are inherently asynchronous. We are currently investigating how to execute these functions synchronously, as this is a requirement for our project, but haven't yet found a solution.~~
~~I was able to build with JSPI and call asynchronous functions synchronously in Chrome. However, JSPI only works in Chrome and Edge 137.~~
~~https://webassembly.org/features/~~
~~https://learn.microsoft.com/en-us/microsoft-edge/web-platform/release-notes/137#javascript-promise-integration-jspi-in-webassembly~~

Supported browsers:

- Chrome v137.0.7151.69
- Firefox ESR v128.11.0esr
- Safari v18.5(20621.2.5.11.8)

shiena commented:
I've tried enabling Asyncify as follows, which I believe should fix it, but the build is still failing.

```diff
diff --git a/platform/web/SCsub b/platform/web/SCsub
index cf872e65ff..5bb105e1cd 100644
--- a/platform/web/SCsub
+++ b/platform/web/SCsub
@@ -106,6 +106,8 @@ else:
     sys_env.Append(LIBS=["idbfs.js"])
     build = sys_env.add_program(build_targets, web_files + ["web_runtime.cpp"])
 
+sys_env.Append(LINKFLAGS=["-sASYNCIFY=1"])
+
 sys_env.Depends(build[0], sys_env["JS_LIBS"])
 sys_env.Depends(build[0], sys_env["JS_PRE"])
 sys_env.Depends(build[0], sys_env["JS_POST"])
```

```
cache:INFO: generating system asset: symbol_lists/8eab473934c10451394fc82e8912c6b3712c9ff9.json... (this will be cached in "C:\Users\shien\dev\emsdk\upstream\emscripten\cache\symbol_lists\8eab473934c10451394fc82e8912c6b3712c9ff9.json" for subsequent builds)
cache:INFO:  - ok
unexpected expression type
UNREACHABLE executed at C:\b\s\w\ir\cache\builder\emscripten-releases\binaryen\src\passes\Asyncify.cpp:1142!
em++: error: 'C:/Users/shien/dev/emsdk/upstream\bin\wasm-opt --strip-target-features --post-emscripten -Os --low-memory-unused --asyncify --pass-arg=asyncify-propagate-addlist --pass-arg=asyncify-imports@env.invoke_*,env.__asyncjs__*,*.fd_sync,*.emscripten_promise_await,*.emscripten_idb_load,*.emscripten_idb_store,*.emscripten_idb_delete,*.emscripten_idb_exists,*.emscripten_idb_clear,*.emscripten_idb_load_blob,*.emscripten_idb_store_blob,*.emscripten_sleep,*.emscripten_wget_data,*.emscripten_scan_registers,*.emscripten_lazy_load_code,*._load_secondary_module,*.emscripten_fiber_swap,*.SDL_Delay --zero-filled-memory --pass-arg=directize-initial-contents-immutable --no-stack-ir bin\godot.web.template_debug.wasm32.nothreads.wasm -o bin\godot.web.template_debug.wasm32.nothreads.wasm -g --mvp-features --enable-bulk-memory --enable-exception-handling --enable-multivalue --enable-mutable-globals --enable-reference-types --enable-sign-ext --enable-simd' failed (returned 3221226505)
scons: *** [bin\godot.web.template_debug.wasm32.nothreads.js] Error 1
```

Faless commented on platform/web/camera_driver_web.cpp:
```diff
@@ -0,0 +1,191 @@
+
+#include <cstdlib>
+
+EM_ASYNC_JS(void, godot_js_camera_get_cameras, (void *context, CameraLibrary_OnGetCamerasCallback p_callback_ptr), {
+	await GodotCamera.api.getCameras(context, p_callback_ptr);
+});
+
+EM_ASYNC_JS(void, godot_js_camera_get_capabilities, (void *context, const char *p_device_id_ptr, CameraLibrary_OnGetCapabilitiesCallback p_callback_ptr), {
+	await GodotCamera.api.getCameraCapabilities(p_device_id_ptr, context, p_callback_ptr);
+});
```
We shouldn't use the `EM_*_JS` functions, as they break dynamic linking.

Faless reviewed (changes requested):
> The demo site requires Chrome 137+ due to its use of JSPI.
> I wanted to use `ASYNCIFY=1` in SCsub but ran into errors, so I'm currently using `ASYNCIFY=2` for JSPI.

This means it can't be merged, we at least want support the latest version of Chrome, Firefox and iOS (and we also try to support the latest version of Firefox ESR, but that has been on and off in the past).

That said, it should be possible to implement the API by relying on callbacks instead of requiring asyncify.

adamscott commented:
Can you share an MRP to test on? I would like to test your PR, but it would be faster for me to have a MRP to test on.

shiena commented:
@adamscott 
This is for all platforms, but I'm testing functionality with this project.
https://github.com/shiena/godot-camerafeed-demo

vonweller commented:
Is this PR not going to be merged?

adamscott commented:
> Is this PR not going to be merged?

We're currently in feature freeze, so it will not be merged for 4.5. We are focused on fixing issues with the upcoming release.

AThousandShips commented on platform/web/camera_driver_web.h:
```diff
@@ -0,0 +1,69 @@
+	static CameraDriverWeb *get_singleton();
+	void get_cameras(void *p_context, CameraDriverWebGetCamerasCallback p_callback);
+	void get_pixel_data(void *p_context, const String &p_device_id, const int p_width, const int p_height, void (*p_callback)(void *, const uint8_t *, const int, const int, const int, const char *), void (*p_denied_callback)(void *));
+	void stop_stream(const String &device_id = String());
```
```suggestion
	void stop_stream(const String &p_device_id = String());
```

AThousandShips commented on modules/camera/camera_web.cpp:
```diff
@@ -0,0 +1,248 @@
+		FeedFormat feed_format;
+		feed_format.width = info.formats[i].width;
+		feed_format.height = info.formats[i].height;
+		feed_format.format = String("RGBA");
```
```suggestion
		feed_format.format = "RGBA";
```
Redundant

shiena committed:
Web: Add CameraFeed support

Implement camera capture for the Web platform using WebCodecs with a
Canvas 2D fallback.

Benjamin-Dobell reviewed (changes requested):
Not a maintainer, but I think there's a few things that will need to be changed. That said, I have tested this and it does indeed work as is, and has unblocked me in the interim, so thank you very much! 🙏

shiena commented on modules/camera/camera_web.cpp:
```diff
@@ -0,0 +1,269 @@
+	CameraServer::set_monitoring_feeds(p_monitoring_feeds);
+	if (p_monitoring_feeds) {
+		if (driver == nullptr) {
+			driver = new CameraDriverWeb();
```
Fixed in 833ccaf38df7fbea00e022ae0c56eaf0d18434fe.

adamscott commented:
I'm gonna take a look at this PR today or tomorrow.

Benjamin-Dobell commented:
@shiena Thanks for making those changes. I'd made some similar changes locally but adopted yours about 24 hours ago. Working great 🙏

Benjamin-Dobell reviewed (approved):
❤️

shiena commented:
A note on the camera frame capture strategy. Browser support for the APIs we'd like to use for pulling pixels out of a `getUserMedia()` stream is uneven, so this PR implements several capture paths and picks the best one available at runtime.

In order of preference:

1. **WebCodecs in a Worker** — `MediaStreamTrackProcessor` produces `VideoFrame`s, which are transferred to a dedicated Worker that extracts RGBA pixels via `VideoFrame.copyTo()`. No canvas, no main-thread pixel work. Used on modern Chromium.

2. **Canvas 2D in a Worker** — `createImageBitmap()` on the `<video>` element produces a transferable bitmap; the Worker draws it to an `OffscreenCanvas` and reads pixels with `getImageData()`. Used when Workers are available but WebCodecs isn't (e.g. current Firefox).

3. **WebCodecs on the main thread** — same `VideoFrame.copyTo()` pipeline as (1), but without offloading. Used when WebCodecs exists but Workers / `OffscreenCanvas` / `createImageBitmap` don't.

4. **Canvas 2D on the main thread** — classic `drawImage(video)` + `getImageData()` driven by `requestAnimationFrame`. Last resort for browsers that lack both of the above (e.g. Safari without WebCodecs).

The tier is chosen once by feature detection when the stream starts. On top of that, each tier also falls back to a lower tier at runtime if its API unexpectedly throws (e.g. `VideoFrame.copyTo()` can fail even when the API is present), so a misdetected environment degrades gracefully instead of breaking the feed.

shiena commented:
Updated the Web CameraFeed implementation:

- Uses WebCodecs + MediaStreamTrackProcessor to read camera frames into CameraFeed textures, with Canvas2D as the fallback path.
- Keeps the initial format list empty until the camera is activated, since browsers may not expose reliable per-device formats before opening the stream.
- Updates CameraFeed formats from the actual stream settings after activation, so the UI reflects the real width, height, and frame rate reported by the browser.
- Removes problematic preset resolutions from the Web format candidates where browser-reported behavior is misleading or unreliable.

shiena commented:
Updated the Web CameraFeed implementation:

- Adds I420/I420A support that feeds the planar YUV frames straight into set_ycbcr_images by interleaving the U/V planes into a single Cb/Cr image and reusing the existing NV12 path, avoiding a per-frame RGBA color conversion (the I420A
alpha plane is ignored).
- Fixes NV12 decoding for odd capture sizes by rounding the chroma plane dimensions up, so streams whose width or height is odd are no longer mis-sized.
- Requests real sensor modes with resizeMode: 'none' and reworks the resolution candidates (dropping monitor-centric presets, adding common 16:9/16:10/4:3 sizes, and sorting them by resolution), so delivered frames use native hardware
modes instead of cropped or scaled ones that often run at a lower frame rate.
- Hardens camera switching on Android, where opening a camera right after stopping another can stall while the previous device is still being released: lengthens the teardown barrier, raises the getUserMedia timeout above libwebrtc's
open-retry budget so a busy device returns a clean AbortError, and retries on that transient error.

shiena commented:
There's one issue on Android Firefox that I haven't been able to work around. Camera switching is problematic because the device close happens asynchronously with no JS callback to signal completion. Opening a new camera too soon after closing the previous one can cause errors or, worse, leave the system in an unrecoverable stuck state. I added retry logic for the error case, but when it gets stuck, the only way out is closing or killing the browser tab.

shiena committed:
Web: Stabilize camera feed activation

Track asynchronous camera feed activation and report activation failures.

On Web, serialize camera operations, cancel stale getUserMedia requests, and discard streams created by canceled startups so format changes and stop/start cycles can restart camera feeds reliably.

Share camera startup logic between activation paths and avoid exposing feed-switching APIs that are not used by the Web camera workflow.

shiena commented:
Updated the Web CameraFeed implementation:
  * Resolves the Android Firefox camera switching issue mentioned above by avoiding the open-before-close camera switch path. Switching now relies on stopping the current stream first, then starting the target stream after the browser
  has had time to release the previous device.
  * Keeps camera activation asynchronous and reports startup failures through `CameraFeed.activation_failed`, while serializing camera operations and canceling stale `getUserMedia()` requests so stop/start and format changes can recover
  cleanly.
  * Updated the demo to use one Web switching strategy: stop the current feed, wait briefly, start the target feed, retry transient startup failures up to 6 times, and fall back to the previous feed if the retry limit is reached.

With this approach, the Android Firefox stuck state is avoided in my testing, including switches involving telephoto/macro cameras.
https://github.com/shiena/godot-camerafeed-demo/tree/feature/web-camera-switch-retry

AThousandShips commented on modules/camera/camera_web.cpp:
```diff
@@ -0,0 +1,167 @@
+	if (error) {
+		String error_str = String::utf8(error);
+		ERR_PRINT(vformat("Camera feed error from JS: %s", error_str));
+		return;
```
```suggestion
		ERR_FAIL_MSG(vformat("Camera feed error from JS: %s.", error_str));
```

AThousandShips commented on modules/camera/camera_web.cpp:
```diff
@@ -0,0 +1,167 @@
+
+	if (context == nullptr || rawdata == nullptr || length < 0 || p_width <= 0 || p_height <= 0) {
+		ERR_PRINT("Camera feed error: Invalid pixel data received.");
+		return;
```
```suggestion
		ERR_FAIL_MSG("Camera feed error: Invalid pixel data received.");
```

AThousandShips commented on modules/camera/camera_web.h:
```diff
@@ -0,0 +1,74 @@
+
+protected:
+public:
+	void set_monitoring_feeds(bool p_monitoring_feeds) override;
```
```suggestion
	virtual void set_monitoring_feeds(bool p_monitoring_feeds) override;
```

AThousandShips commented on modules/camera/camera_web.h:
```diff
@@ -0,0 +1,74 @@
+
+protected:
+public:
+	bool activate_feed() override;
```
```suggestion
	virtual bool activate_feed() override;
```
For all these too

AThousandShips commented on doc/classes/CameraServer.xml:
```diff
@@ -6,7 +6,7 @@
 		The [CameraServer] keeps track of different cameras accessible in Godot. These are external cameras such as webcams or the cameras on your phone.
 		It is notably used to provide AR modules with a video feed from the camera.
-		[b]Note:[/b] This class is currently only implemented on Linux, Android, macOS, and iOS. On other platforms no [CameraFeed]s will be available. To get a [CameraFeed] on iOS, enable [member EditorExportPlatformIOS.modules/camera].
+		[b]Note:[/b] This class is currently only implemented on Linux, Android, macOS, iOS and Web. On other platforms no [CameraFeed]s will be available. To get a [CameraFeed] on iOS, enable [member EditorExportPlatformIOS.modules/camera].
```
```suggestion
		[b]Note:[/b] This class is currently only implemented on Linux, Android, macOS, iOS, and web. On other platforms no [CameraFeed]s will be available. To get a [CameraFeed] on iOS, enable [member EditorExportPlatformIOS.modules/camera].
```

AThousandShips commented on doc/classes/CameraFeed.xml:
```diff
@@ -6,7 +6,7 @@
 		A camera feed gives you access to a single physical camera attached to your device. When enabled, Godot will start capturing frames from the camera which can then be used. See also [CameraServer].
 		[b]Note:[/b] Many cameras will return YCbCr images which are split into two textures and need to be combined in a shader. Godot does this automatically for you if you set the environment to show the camera image in the background.
-		[b]Note:[/b] This class is currently only implemented on Linux, Android, macOS, and iOS. On other platforms no [CameraFeed]s will be available. To get a [CameraFeed] on iOS, enable [member EditorExportPlatformIOS.modules/camera].
+		[b]Note:[/b] This class is currently only implemented on Linux, Android, macOS, iOS and Web. On other platforms no [CameraFeed]s will be available. To get a [CameraFeed] on iOS, enable [member EditorExportPlatformIOS.modules/camera].
```
```suggestion
		[b]Note:[/b] This class is currently only implemented on Linux, Android, macOS, iOS, and web. On other platforms no [CameraFeed]s will be available. To get a [CameraFeed] on iOS, enable [member EditorExportPlatformIOS.modules/camera].
```

AThousandShips commented on modules/camera/camera_web.cpp:
```diff
@@ -0,0 +1,330 @@
+		default:
+			// Unknown format - drop frames to avoid mis-sizing the destination image.
+			ERR_PRINT(vformat("Camera feed error: Unknown pixel format code %d.", p_pixel_format));
+			return memnew(NullBufferDecoder(p_feed));
```
```suggestion
			ERR_FAIL_V_MSG(memnew(NullBufferDecoder(p_feed)), vformat("Camera feed error: Unknown pixel format code %d.", p_pixel_format));
```

AThousandShips commented on platform/web/camera_driver_web.cpp:
```diff
@@ -0,0 +1,137 @@
+	Array devices_array = v_devices;
+	Vector<CameraInfo> camera_info;
+	for (int i = 0; i < devices_array.size(); i++) {
+		Variant device_variant = devices_array.get(i);
```
```suggestion
	for (Variant device_variant : devices_array) {
```

AThousandShips commented on modules/camera/camera_web.cpp:
```diff
@@ -0,0 +1,464 @@
+	Vector<FeedFormat> new_formats;
+	Array formats_array = v_formats;
+	for (int i = 0; i < formats_array.size(); i++) {
+		Variant format_variant = formats_array[i];
```
```suggestion
	for (Variant format_variant : formats_array) {
```

AThousandShips commented on modules/camera/camera_web.cpp:
```diff
@@ -0,0 +1,464 @@
+
+bool CameraFeedWeb::set_format(int p_index, const Dictionary &p_parameters) {
+	(void)p_parameters;
+
```
```suggestion
```

AThousandShips commented on modules/camera/camera_web.cpp:
```diff
@@ -0,0 +1,464 @@
+	// Build a set of new device IDs for quick lookup.
+	HashSet<String> new_device_ids;
+	for (int i = 0; i < p_camera_info.size(); i++) {
+		new_device_ids.insert(p_camera_info[i].device_id);
```
```suggestion
	for (const CameraInfo &camera_info : p_camera_info) {
		new_device_ids.insert(camera_info.device_id);
```

AThousandShips commented on modules/camera/camera_web.cpp:
```diff
@@ -0,0 +1,464 @@
+	// Build a set of existing device IDs.
+	HashSet<String> existing_device_ids;
+	for (int i = 0; i < server->feeds.size(); i++) {
+		Ref<CameraFeedWeb> feed = server->feeds[i];
```
```suggestion
	for (Ref<CameraFeedWeb> feed : server->feeds) {
```

AThousandShips commented on modules/camera/camera_web.cpp:
```diff
@@ -0,0 +1,464 @@
+	// Add new feeds that don't already exist.
+	for (int i = 0; i < p_camera_info.size(); i++) {
+		if (!existing_device_ids.has(p_camera_info[i].device_id)) {
+			Ref<CameraFeedWeb> feed = memnew(CameraFeedWeb(p_camera_info[i]));
```
```suggestion
	for (const CameraInfo &camera_info : p_camera_info) {
		if (!existing_device_ids.has(camera_info.device_id)) {
			Ref<CameraFeedWeb> feed = memnew(CameraFeedWeb(camera_info));
```

AThousandShips commented on modules/camera/camera_web.cpp:
```diff
@@ -0,0 +1,464 @@
+			feed->call_deferred("emit_signal", SNAME("format_changed"));
+		} else {
+			for (int i = 0; i < feed->formats.size(); i++) {
+				feed->formats.write[i].format = feed->detected_format_name;
```
```suggestion
			for (FeedFormat &feed_format : feed->formats) {
				feed_format.format = feed->detected_format_name;
```

shiena committed:
Apply review

shiena commented:
Thank you for the review. Applied all suggestions.
